src/Featurizers/AnalyticalRollingWindowFeaturizer.h

```diff
@@ -0,0 +1,307 @@
+    bool operator!=(AnalyticalRollingWindowTransformer const &other) const;
+    void save(Archive & ar) const override;
+
+private:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Types
+    // |
+    // ----------------------------------------------------------------------
+    using OutputType = std::vector<double>;    
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Members
+    // |
+    // ----------------------------------------------------------------------
+    // _windowSizes and _minWindowCount are int64 so that we can have the full positive range of the uint32 it comes
+    // in as but can still support negative numbers.
+    Components::CircularBuffer<InputT>      _buffer;
+    std::int64_t                            _windowSize;
+    std::uint32_t                           _horizon;
+    std::int64_t                            _minWindowCount;
+    AnalyticalRollingWindowCalculation      _windowCalculation;
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Methods
+    // |
+    // ----------------------------------------------------------------------
+
+    // MSVC runs into problems when the declaration and definition are separated
+    void execute_impl(typename BaseType::InputType const &input, typename BaseType::CallbackFunction const &callback) override {
+        OutputType         results(_horizon);
+        
+        _buffer.push(input);
+        size_t bufferSize = _buffer.size();
+
+        for (int i = _horizon; i > 0; i--){
+            // We cast to int32 here because -1 is valid for the algorithm.
+            std::int32_t elementsAvailableToUse(static_cast<std::int32_t>(bufferSize) - i);
+
+            // If we don't have enough elements then output NaN
+            if (elementsAvailableToUse < _minWindowCount) {
+                results[_horizon - i] = std::nan("");
+
+            // If we have strictly less elements then the window size, but more then the minimum size
+            } else if (elementsAvailableToUse < _windowSize) {
+                const auto range = _buffer.range(_windowSize - elementsAvailableToUse);
+                auto start_iter = std::get<0>(range);
+                auto end_iter = std::get<1>(range);
+
+                results[_horizon - i] = Calculators::MeanCalculator<BaseType::InputType>::execute(start_iter, end_iter);
+
+            // More elements are in the buffer than are needed in the window
+            } else {
+                const auto range = _buffer.range(_windowSize, bufferSize - i - _windowSize);
+                auto start_iter = std::get<0>(range);
+                auto end_iter = std::get<1>(range);
+
+                results[_horizon - i] = Calculators::MeanCalculator<BaseType::InputType>::execute(start_iter, end_iter);
+            }

```

Comment:
It looks like the only difference between these 2 else blocks is how the range is calculated. Merge these 2 into a single else clause and use a ternary statement or a lambda function to properly calculate the range.

Reply:
Resolving this based on our offline discussion.